rasa/utils/tensorflow/models.py

```diff
@@ -187,10 +187,31 @@ def train_on_batch(
     ) -> None:
         """Train on batch"""
 
-        with tf.GradientTape() as tape:
-            total_loss = self._total_batch_loss(batch_in)
+        with tf.GradientTape(persistent=True) as tape:
+            prediction_loss = self.batch_loss(batch_in)
+            regularization_loss = tf.math.add_n(self.losses)
+            total_loss = prediction_loss + regularization_loss
+
+        self.total_loss.update_state(total_loss)
+
+        prediction_gradients = tape.gradient(prediction_loss, self.trainable_variables)
+        regularization_gradients = tape.gradient(
+            regularization_loss, self.trainable_variables
+        )
+        del tape
```

Comment:
Why do we delete it?

Reply:
that's what they say in tf docs, because we create it with `persistent=True`

Reply:
added comment